File: mobile/apps/photos_tv/lib/src/pairing_service.dart
import 'dart:convert';

import 'package:ente_crypto_api/ente_crypto_api.dart';
import 'package:http/http.dart' as http;

import 'config.dart';
import 'http_util.dart';
import 'models.dart';

/// Service that registers this TV and fetches cast payloads.
class PairingService {
  final http.Client _client;

  /// Creates pairing service.
  PairingService(this._client);

  /// Registers this TV for pairing.
  Future<Registration> register() async {
    final keyPair = CryptoUtil.generateKeyPair();
    final publicKey = CryptoUtil.bin2base64(keyPair.publicKey);
    final pairingCode = await _registerDevice(publicKey);
    return Registration(
      pairingCode: pairingCode,
      publicKey: publicKey,
      privateKey: CryptoUtil.bin2base64(keyPair.secretKey),
    );
  }

  /// Fetches decrypted cast payload for registration.
  Future<CastPayload?> getCastPayload(Registration registration) async {
    final encryptedCastData = await _getEncryptedCastData(
      registration.pairingCode,
    );
    if (encryptedCastData == null) return null;
    final payloadBytes = CryptoUtil.openSealSync(
      CryptoUtil.base642bin(encryptedCastData),
      CryptoUtil.base642bin(registration.publicKey),
      CryptoUtil.base642bin(registration.privateKey),
    );
    return CastPayload.fromJson(jsonDecode(utf8.decode(payloadBytes)));
  }

  /// Closes HTTP resources.
  void close() => _client.close();

  Future<String> _registerDevice(String publicKey) async {
    final response = await _client.post(
      Uri.parse('$apiOrigin/cast/device-info'),
      headers: jsonHeaders,
      body: jsonEncode({'publicKey': publicKey}),
    );
    ensureOk(response);
    return jsonDecode(response.body)['deviceCode'] as String;
  }

  Future<String?> _getEncryptedCastData(String code) async {
    final response = await _client.get(
      Uri.parse('$apiOrigin/cast/cast-data/$code'),
      headers: jsonHeaders,
    );
    ensureOk(response);
    return jsonDecode(response.body)['encCastData'] as String?;
  }
}

File: mobile/apps/photos_tv/android/app/src/main/kotlin/io/ente/photos_tv/PairingService.kt
package io.ente.photos_tv

import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.put
import okhttp3.OkHttpClient
import okhttp3.Request

internal class PairingService(
    private val client: OkHttpClient,
    private val cryptoBox: CryptoBox,
) {
    fun register(): Registration {
        val keyPair = cryptoBox.generateKeyPair()
        val publicKey = cryptoBox.base64(keyPair.publicKey)
        val pairingCode = registerDevice(publicKey)
        return Registration(
            pairingCode = pairingCode,
            publicKey = publicKey,
            privateKey = cryptoBox.base64(keyPair.privateKey),
        )
    }

    fun getCastPayload(registration: Registration): CastPayload? {
        val encryptedCastData = getEncryptedCastData(registration.pairingCode) ?: return null
        val payloadBytes = cryptoBox.openSeal(
            input = cryptoBox.base64Decode(encryptedCastData),
            publicKey = cryptoBox.base64Decode(registration.publicKey),
            privateKey = cryptoBox.base64Decode(registration.privateKey),
        )
        return JsonConfig.value.decodeFromString<CastPayload>(payloadBytes.decodeToString())
    }

    private fun registerDevice(publicKey: String): String {
        val body = buildJsonObject {
            put("publicKey", publicKey)
        }.toString().toJsonRequestBody()
        val request = Request.Builder()
            .url("$API_ORIGIN/cast/device-info")
            .jsonHeaders()
            .post(body)
            .build()
        client.newCall(request).execute().use { response ->
            response.ensureOk()
            return JsonConfig.value.parseToJsonElement(response.body!!.string()).jsonObject.getString("deviceCode")
        }
    }

    private fun getEncryptedCastData(code: String): String? {
        val request = Request.Builder()
            .url("$API_ORIGIN/cast/cast-data/$code")
            .jsonHeaders()
            .get()
            .build()
        client.newCall(request).execute().use { response ->
            response.ensureOk()
            return JsonConfig.value.parseToJsonElement(response.body!!.string()).jsonObject["encCastData"]?.jsonPrimitive?.contentOrNull
        }
    }
}
